daemon/create_windows.go

```diff
@@ -80,3 +85,67 @@ func (daemon *Daemon) createContainerOSSpecificSettings(container *container.Con
 	}
 	return nil
 }
+
+// checkImageIsolationCompatiblity checks whether the provided image is compatible
+// with the host config's isolation settings.
+func (daemon *Daemon) checkImageIsolationCompatiblity(hostConfig *containertypes.HostConfig, img *image.Image) error {
+	hostOSV := osversion.Get()
+	isHyperV := hostConfig.Isolation.IsHyperV()
+	// fallback on default daemon isolation if none was explicitly provided:
+	if hostConfig.Isolation.IsDefault() {
+		isHyperV = daemon.defaultIsolation.IsHyperV()
+	}
+	return checkImageCompatibilityForHostIsolation(hostOSV, img.OSVersion, isHyperV)
+}
+
+// checkImageCompatibilityForHostIsolation checks whether the provided `imageOSVersion`
+// should be able to run on the provided host OS version given Hyper-V isolation.
+// Its contained logic can be distilled into:
+// - if imageOS == hostOS => can run under any isolation mode
+// - if imageOS < hostOS => must be run under Hyper-V
+// - if imageOS > hostOS => can be run under Hyper-V only on an RS5 (1809+) host
+// Please see the below:
+// https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2016%2Cwindows-10#windows-server-host-os-compatibility
+func checkImageCompatibilityForHostIsolation(hostOSV osversion.OSVersion, imageOSVersion string, isHyperV bool) error {
```

Comment:
Does this break any existing behavior? It sounds like if we allow pulling any image version on a host >= RS5 we unblock Win10 hosts. These checks have the potential to break something else that is currently working even though it would be contrary to the documentation.

Reply:
> These checks have the potential to break something else that is currently working even though it would be contrary to the documentation.

Technically yes, the removal of the image filtering on RS5+ during image pulling could allow someone to pull a newer image which would only be able to run under Hyper-V, in which case they'd get an error from HCS itself if they tried running it under Process isolation.

Note that this potential error was already there if `imageOS < hostOS` (which requires Hyper-V isolation), and [my removing the pulling check newer images for hosts RS5+](https://github.com/moby/moby/pull/44489/files#diff-6019c68f8c9392b97b32ce83ecaa765fdd517946fc9685ba523182f647cd2981R131) will just mean this same error might show up if `imageOS > hostOS` too.

Considering we can't know under what isolation the user plans to run the image during pulling time (outside of maybe `docker run` commands which automatically pull the image the first time if it's not present), it is **IMO better to allow users to pull images and have Moby/HCS tell them if the image is compatible** with the chosen isolation mode or give a clear error message if not. (which is what I was trying to do here in `checkImageCompatibilityForHostIsolation`)

It's also worth noting there are afaik no tests or docs for the logic introduced in #36327 which prevents the image pulling, nor could I find any mention of newer image pulling being prevented in the docs. (all references seem to point to [the compatibility matrix in Microsoft's docs](https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2016%2Cwindows-10#windows-server-host-os-compatibility))